planner/core/resolve_indices.go

```diff
@@ -632,13 +635,60 @@ func (p *baseSchemaProducer) ResolveIndices() (err error) {
 	return
 }
 
+// ResolveIndices implements Plan interface.
+func (p *basePhysicalJoin) ResolveIndices() (err error) {
+	for _, child := range p.children {
+		err = child.ResolveIndices()
+		if err != nil {
+			return err
+		}
+	}
+	switch p.JoinType {
+	case RightOuterJoin:
+		p.AlignNullFlag(p.children[1].Schema())
+		p.EnsureNullFlag(p.schema, 0, len(p.children[0].Schema().Columns))
+	case LeftOuterJoin:
+		p.AlignNullFlag(p.children[0].Schema())
+		p.EnsureNullFlag(p.schema, len(p.children[0].Schema().Columns), len(p.schema.Columns))
+	default:
+		p.AlignNullFlag(p.children[0].Schema())
+		p.AlignNullFlag(p.children[1].Schema())
+	}
+	return
+}
+
 // ResolveIndices implements Plan interface.
 func (p *basePhysicalPlan) ResolveIndices() (err error) {
 	for _, child := range p.children {
 		err = child.ResolveIndices()
 		if err != nil {
 			return err
 		}
+		p.AlignNullFlag(child.Schema())
 	}
 	return
 }
+
+// AlignNullFlag aligns the parent's column not null flag to its child not null flag when these columns uniqueID are identical.
+// because before outer joins are optimized into inner joins, original inner columns in outer joins are set nullable, we should recover the original flag if the optimization succeeds.
+func (p *basePhysicalPlan) AlignNullFlag(schema *expression.Schema) {
+	for _, pcol := range p.self.Schema().Columns {
+		for _, scol := range schema.Columns {
+			if pcol.UniqueID == scol.UniqueID {
+				if pcol.RetType.Flag != scol.RetType.Flag {
```

Comment:
According to the function name and comment, we should only modify the NotNull flag here, instead of the whole flag.